File: src/widgets/cwd.zig
const std = @import("std");
const tmux_renderer = @import("../tmux_renderer.zig");
const WidgetContext = @import("../core/widget.zig").WidgetContext;

pub fn run(
    allocator: std.mem.Allocator,
    io: std.Io,
    theme_name: []const u8,
    transparent: bool,
    environ_map: *std.process.Environ.Map,
    cwd: []const u8,
    writer: *std.Io.Writer,
) !void {
    var ctx = try WidgetContext.init(allocator, io, environ_map, theme_name, transparent);
    defer ctx.deinit();
    const theme = ctx.theme;
    const reset = ctx.reset;

    const path = if (cwd.len > 0) cwd else return;

    // Get the basename for display
    const basename = std.fs.path.basename(path);
    if (basename.len == 0) return;

    // Try to get git repo root for relative display
    const git_result = std.process.run(allocator, io, .{
        .argv = &.{
            "git", "rev-parse", "--show-toplevel",
        },
        .cwd = .{ .path = path },
    }) catch null;

    var display_path: []const u8 = basename;
    var display_owned: bool = false;

    if (git_result) |result| {
        defer allocator.free(result.stderr);
        defer allocator.free(result.stdout);
        if (result.term == .exited and result.term.exited == 0) {
            const repo_root = std.mem.trim(u8, result.stdout, " \n\r\t");
            if (repo_root.len > 0) {
                const repo_name = std.fs.path.basename(repo_root);
                if (repo_name.len > 0) {
                    if (std.mem.eql(u8, repo_name, basename)) {
                        display_path = basename;
                    } else {
                        const rel = try std.fmt.allocPrint(allocator, "{s}/{s}", .{ repo_name, basename });
                        display_path = rel;
                        display_owned = true;
                    }
                }
            }
        }
    }
    defer if (display_owned) allocator.free(display_path);

    var emphasis_hex: [32]u8 = undefined;
    var bg_hex: [32]u8 = undefined;
    const emphasis_str = tmux_renderer.colorHexString(theme.emphasis, &emphasis_hex);
    const bg_str = tmux_renderer.colorHexString(theme.background, &bg_hex);
    try writer.print("{s}#[fg={s},bg={s},bold]󰉋 {s}", .{
        reset,
        emphasis_str,
        bg_str,
        display_path,
    });
}

test "cwd WidgetContext initializes" {
    const gpa = std.testing.allocator;
    const io = std.Io.threaded_global.ioBasic();
    var env_map = std.process.Environ.Map.init(gpa);
    defer env_map.deinit();

    var ctx = try WidgetContext.init(gpa, io, &env_map, "hard", false);
    defer ctx.deinit();

    try std.testing.expect(!ctx.theme.emphasis.isDefault());
    try std.testing.expect(!ctx.theme.background.isDefault());
    try std.testing.expect(std.mem.startsWith(u8, ctx.reset, "#[fg="));
}

test "cwd produces output for tmp dir" {
    const gpa = std.testing.allocator;
    const io = std.Io.threaded_global.ioBasic();
    var env_map = std.process.Environ.Map.init(gpa);
    defer env_map.deinit();

    var buf: [2048]u8 = undefined;
    var writer: std.Io.Writer = .fixed(&buf);

    run(gpa, io, "hard", false, &env_map, "/tmp", &writer) catch return error.SkipZigTest;
    const output = std.Io.Writer.buffered(&writer);

    try std.testing.expect(output.len > 0);
    try std.testing.expect(std.mem.containsAtLeast(u8, output, 1, "#[fg="));
    try std.testing.expect(std.mem.containsAtLeast(u8, output, 1, "󰉋"));
}

File: src/widgets/cwd.rs
use std::collections::hash_map::DefaultHasher;
use std::fs;
use std::hash::Hasher;
use std::path::{Path, PathBuf};
use std::process::Command;

use crate::core::cache::{cache_base, read_fresh_cache, write_cache};
use crate::core::widget::WidgetContext;
use crate::core::Theme;
use crate::tmux_renderer::ThemeHex;

const CACHE_TTL_SECONDS: u64 = 2;

fn absolute_path(path: &Path) -> PathBuf {
    if let Ok(canonical) = fs::canonicalize(path) {
        return canonical;
    }
    if path.is_absolute() {
        return path.to_path_buf();
    }
    match std::env::current_dir() {
        Ok(current) => current.join(path),
        Err(_) => path.to_path_buf(),
    }
}

fn cache_path(absolute_cwd: &Path) -> String {
    let mut hasher = DefaultHasher::new();
    hasher.write(absolute_cwd.to_string_lossy().as_bytes());
    let key = hasher.finish();

    format!("{}/flavors-tmux-cwd-{:x}.cache", cache_base(), key)
}

fn git_repo_root(cwd: &Path) -> String {
    match Command::new("git")
        .args(["rev-parse", "--show-toplevel"])
        .current_dir(cwd)
        .output()
    {
        Ok(output) if output.status.success() => {
            String::from_utf8(output.stdout).unwrap_or_default().trim().to_owned()
        }
        _ => String::new(),
    }
}

fn cached_git_repo_root(absolute_cwd: &Path) -> String {
    let path = cache_path(absolute_cwd);
    if let Some(cached) = read_fresh_cache(&path, CACHE_TTL_SECONDS) {
        return cached;
    }

    let repo_root = git_repo_root(absolute_cwd);
    write_cache(&path, &repo_root);
    repo_root
}

fn display_path(basename: String, repo_root: &str) -> String {
    if repo_root.is_empty() {
        return basename;
    }

    let repo_name = Path::new(repo_root)
        .file_name()
        .map(|n| n.to_string_lossy().into_owned())
        .unwrap_or_default();
    if repo_name.is_empty() || repo_name == basename {
        basename
    } else {
        format!("{}/{}", repo_name, basename)
    }
}

/// Render the current working directory widget.
/// Shows `󰉋 basename` outside a git repo, or `󰉋 repo-name/basename` inside one.
pub fn run(theme: Theme, cwd: &str) -> String {
    let theme_hex = ThemeHex::from_theme(theme);
    run_with_theme_hex(theme, &theme_hex, cwd)
}

pub(crate) fn run_with_theme_hex(theme: Theme, theme_hex: &ThemeHex, cwd: &str) -> String {
    let ctx = WidgetContext::from_theme_hex(theme, theme_hex);
    let theme = ctx.theme;

    if cwd.is_empty() {
        return String::new();
    }

    let path = Path::new(cwd);
    let basename = match path.file_name() {
        Some(name) => name.to_string_lossy().into_owned(),
        None => return String::new(),
    };

    let absolute_cwd = absolute_path(path);
    let repo_root = cached_git_repo_root(&absolute_cwd);
    let display_path = display_path(basename, &repo_root);

    format!(
        "{}#[fg={},bg={},bold]󰉋 {}",
        ctx.reset,
        theme_hex.color(theme.emphasis),
        theme_hex.color(theme.surface),
        display_path,
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::themes;

    #[test]
    fn cwd_produces_output_for_tmp_dir() {
        let output = run(themes::hard::THEME, "/tmp");
        assert!(output.contains("#[fg="));
        assert!(output.contains("󰉋"));
    }

    #[test]
    fn display_path_uses_basename_when_repo_root_is_empty() {
        let output = display_path(String::from("project"), "");

        assert_eq!(output, "project");
    }

    #[test]
    fn display_path_prefixes_repo_name_when_inside_repo_subdir() {
        let output = display_path(String::from("src"), "/work/project");

        assert_eq!(output, "project/src");
    }

    #[test]
    fn run_uses_fresh_cached_repo_root_when_present() {
        let unique = std::time::SystemTime::now()
            .duration_since(std::time::SystemTime::UNIX_EPOCH)
            .unwrap()
            .as_nanos();
        let cwd = std::env::temp_dir().join(format!("flavors-tmux-cwd-test-{unique}"));
        std::fs::create_dir_all(&cwd).unwrap();
        let absolute_cwd = absolute_path(&cwd);
        let cache = cache_path(&absolute_cwd);
        write_cache(&cache, "/work/fake-repo");

        let output = run(themes::hard::THEME, &cwd.to_string_lossy());

        assert!(output.contains("fake-repo/"));
        std::fs::remove_file(cache).unwrap();
        std::fs::remove_dir_all(cwd).unwrap();
    }
}
